File: web/js/src/FFmpegKit.js
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

import {parseArguments} from './Arguments.js';
import {FFmpegKitConfig} from './FFmpegKitConfig.js';
import {getFactory} from './FFmpegKitFactory.js';
import {FFmpegSession} from './FFmpegSession.js';

/**
 * Main entry point for running FFmpeg commands in the web worker.
 *
 * Commands execute inside the FFmpegKit worker. Methods that accept a command
 * string parse it into arguments using FFmpegKitConfig.parseArguments(); use
 * the argument-array methods when arguments must be passed losslessly.
 *
 * Every method here is async, as on the native platforms, so a bad argument rejects
 * the returned promise instead of throwing out of the call and slipping past a
 * caller's .catch().
 */
export class FFmpegKit {
    /**
     * Runs an FFmpeg command to completion.
     *
     * On web the public promise waits for completion, while native FFmpeg runs
     * asynchronously inside the worker so cancel and live-progress messages can
     * still be processed.
     *
     * @param {string} command Command string to parse and execute.
     * @returns {Promise<FFmpegSession>} a populated session after the command reaches COMPLETED or FAILED.
     */
    static async execute(command) {
        return FFmpegKit.executeWithArguments(parseArguments(command));
    }

    /**
     * Runs an FFmpeg command to completion using pre-split arguments.
     *
     * On web the public promise waits for completion, while native FFmpeg runs
     * asynchronously inside the worker so cancel and live-progress messages can
     * still be processed.
     *
     * @param {string[]} commandArguments Command arguments passed without string parsing.
     * @returns {Promise<FFmpegSession>} a populated session after the command reaches COMPLETED or FAILED.
     */
    static async executeWithArguments(commandArguments) {
        const session = await FFmpegSession.create(commandArguments);
        await FFmpegKitConfig.ffmpegExecute(session);
        return session;
    }

    /**
     * Starts an asynchronous FFmpeg command.
     *
     * The promise resolves with the session after the request is posted to the worker;
     * it does not wait for completion. Use completeCallback to observe the final
     * populated session. logCallback and statisticsCallback receive live events
     * while the command runs.
     *
     * @param {string} command Command string to parse and execute.
     * @param {?FFmpegSessionCompleteCallback} completeCallback Optional completion callback.
     * @param {?LogCallback} logCallback Optional per-session log callback.
     * @param {?StatisticsCallback} statisticsCallback Optional per-session statistics callback.
     * @returns {Promise<FFmpegSession>} the session after the request is posted to the worker.
     */
    static async executeAsync(
        command,
        completeCallback = null,
        logCallback = null,
        statisticsCallback = null
    ) {
        return FFmpegKit.executeWithArgumentsAsync(
            parseArguments(command),
            completeCallback,
            logCallback,
            statisticsCallback
        );
    }

    /**
     * Starts an asynchronous FFmpeg command using pre-split arguments.
     *
     * The promise resolves with the session after the request is posted to the worker;
     * it does not wait for completion.
     *
     * @param {string[]} commandArguments Command arguments passed without string parsing.
     * @param {?FFmpegSessionCompleteCallback} completeCallback Optional completion callback.
     * @param {?LogCallback} logCallback Optional per-session log callback.
     * @param {?StatisticsCallback} statisticsCallback Optional per-session statistics callback.
     * @returns {Promise<FFmpegSession>} the session after the request is posted to the worker.
     */
    static async executeWithArgumentsAsync(
        commandArguments,
        completeCallback = null,
        logCallback = null,
        statisticsCallback = null
    ) {
        const session = await FFmpegSession.create(
            commandArguments,
            completeCallback,
            logCallback,
            statisticsCallback
        );
        await FFmpegKitConfig.asyncFFmpegExecute(session);
        return session;
    }

    /**
     * Cancels a running session, or all ongoing sessions when no session id is
     * provided.
     *
     * On web, cancel requests are processed by the worker event loop and can
     * target both execute/executeWithArguments and executeAsync/executeWithArgumentsAsync
     * while native FFmpeg is still running.
     *
     * @param {?number} sessionId Optional session id to cancel.
     * @returns {Promise<void>} resolves after the cancel request is sent.
     */
    static async cancel(sessionId = null) {
        return getFactory().cancel(sessionId);
    }

    /**
     * @returns {Promise<FFmpegSession[]>} all FFmpeg sessions still in the native session history.
     */
    static async listSessions() {
        return getFactory().listFFmpegSessions();
    }
}

File: apple/src/FFmpegKit.m
/*
 * Copyright (c) 2018-2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#import "fftools/ffmpeg.h"
#import "ArchDetect.h"
#import "AtomicLong.h"
#import "FFmpegKit.h"
#import "FFmpegKitConfig.h"
#import "Packages.h"

@implementation FFmpegKit

+ (void)initialize {
    [FFmpegKitConfig class];

    #if TARGET_OS_MACCATALYST
        NSString* osType = @"maccatalyst";
    #elif TARGET_OS_VISION
        NSString* osType = @"visionos";
    #elif TARGET_OS_TV
        NSString* osType = @"tvos";
    #elif TARGET_OS_IOS
        NSString* osType = @"ios";
    #elif TARGET_OS_MAC
        NSString* osType = @"macos";
    #endif

    NSString* packageName = [Packages getPackageName];
    NSString* packageNamePart = [packageName length] > 0 ? [NSString stringWithFormat:@"%@-", packageName] : @"";

    NSLog(@"Loaded ffmpeg-kit-next-%@%@-%@-%@%@-%@.\n", packageNamePart,
          [ArchDetect getArch], [FFmpegKitConfig getVersion], osType,
          [ArchDetect getMinSdk], [FFmpegKitConfig getBuildDate]);
}

+ (FFmpegSession *)executeWithArguments:(NSArray *)arguments {
    FFmpegSession *session = [FFmpegSession create:arguments];
    [FFmpegKitConfig ffmpegExecute:session];
    return session;
}

+ (FFmpegSession *)executeWithArgumentsAsync:(NSArray *)arguments
                        withCompleteCallback:
                            (FFmpegSessionCompleteCallback)completeCallback {
    FFmpegSession *session = [FFmpegSession create:arguments
                              withCompleteCallback:completeCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session];
    return session;
}

+ (FFmpegSession *)
    executeWithArgumentsAsync:(NSArray *)arguments
         withCompleteCallback:(FFmpegSessionCompleteCallback)completeCallback
              withLogCallback:(LogCallback)logCallback
       withStatisticsCallback:(StatisticsCallback)statisticsCallback {
    FFmpegSession *session = [FFmpegSession create:arguments
                              withCompleteCallback:completeCallback
                                   withLogCallback:logCallback
                            withStatisticsCallback:statisticsCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session];
    return session;
}

+ (FFmpegSession *)executeWithArgumentsAsync:(NSArray *)arguments
                        withCompleteCallback:
                            (FFmpegSessionCompleteCallback)completeCallback
                             onDispatchQueue:(dispatch_queue_t)queue {
    FFmpegSession *session = [FFmpegSession create:arguments
                              withCompleteCallback:completeCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session onDispatchQueue:queue];
    return session;
}

+ (FFmpegSession *)
    executeWithArgumentsAsync:(NSArray *)arguments
         withCompleteCallback:(FFmpegSessionCompleteCallback)completeCallback
              withLogCallback:(LogCallback)logCallback
       withStatisticsCallback:(StatisticsCallback)statisticsCallback
              onDispatchQueue:(dispatch_queue_t)queue {
    FFmpegSession *session = [FFmpegSession create:arguments
                              withCompleteCallback:completeCallback
                                   withLogCallback:logCallback
                            withStatisticsCallback:statisticsCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session onDispatchQueue:queue];
    return session;
}

+ (FFmpegSession *)execute:(NSString *)command {
    FFmpegSession *session =
        [FFmpegSession create:[FFmpegKitConfig parseArguments:command]];
    [FFmpegKitConfig ffmpegExecute:session];
    return session;
}

+ (FFmpegSession *)executeAsync:(NSString *)command
           withCompleteCallback:
               (FFmpegSessionCompleteCallback)completeCallback {
    FFmpegSession *session =
        [FFmpegSession create:[FFmpegKitConfig parseArguments:command]
            withCompleteCallback:completeCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session];
    return session;
}

+ (FFmpegSession *)executeAsync:(NSString *)command
           withCompleteCallback:(FFmpegSessionCompleteCallback)completeCallback
                withLogCallback:(LogCallback)logCallback
         withStatisticsCallback:(StatisticsCallback)statisticsCallback {
    FFmpegSession *session =
        [FFmpegSession create:[FFmpegKitConfig parseArguments:command]
              withCompleteCallback:completeCallback
                   withLogCallback:logCallback
            withStatisticsCallback:statisticsCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session];
    return session;
}

+ (FFmpegSession *)executeAsync:(NSString *)command
           withCompleteCallback:(FFmpegSessionCompleteCallback)completeCallback
                onDispatchQueue:(dispatch_queue_t)queue {
    FFmpegSession *session =
        [FFmpegSession create:[FFmpegKitConfig parseArguments:command]
            withCompleteCallback:completeCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session onDispatchQueue:queue];
    return session;
}

+ (FFmpegSession *)executeAsync:(NSString *)command
           withCompleteCallback:(FFmpegSessionCompleteCallback)completeCallback
                withLogCallback:(LogCallback)logCallback
         withStatisticsCallback:(StatisticsCallback)statisticsCallback
                onDispatchQueue:(dispatch_queue_t)queue {
    FFmpegSession *session =
        [FFmpegSession create:[FFmpegKitConfig parseArguments:command]
              withCompleteCallback:completeCallback
                   withLogCallback:logCallback
            withStatisticsCallback:statisticsCallback];
    [FFmpegKitConfig asyncFFmpegExecute:session onDispatchQueue:queue];
    return session;
}

+ (void)cancel {

    /*
     * ZERO (0) IS A SPECIAL SESSION ID
     * WHEN IT IS PASSED TO THIS METHOD, A SIGINT IS GENERATED WHICH CANCELS ALL
     * ONGOING SESSIONS
     */
    cancel_operation(0);
}

+ (void)cancel:(long)sessionId {
    cancel_operation(sessionId);
}

+ (NSArray *)listSessions {
    return [FFmpegKitConfig getFFmpegSessions];
}

@end
